mujoco_warp/_src/io.py

```diff
@@ -1303,6 +1304,547 @@ def reset_contact(
+      dot_prod = float(0.0)
+      for i in range(nv):
+        dot_prod += body_jac_row_in[worldid, i] * result_vec_in[worldid, i]
+      body_A_diag_out[body_A_diag_id, bodyid_target, row_idx] = dot_prod
+
+    @nested_kernel(module="unique", enable_backward=False)
+    def finalize_body_invweight0(
+      body_weldid: wp.array(dtype=int),
+      body_A_diag_in: wp.array3d(dtype=float),
+      body_invweight0_out: wp.array2d(dtype=wp.vec2),
+    ):
+      worldid, bodyid = wp.tid()
+      body_invweight0_id = worldid % body_invweight0_out.shape[0]
+      body_A_diag_id = worldid % body_A_diag_in.shape[0]
+
+      # World body and static bodies have zero invweight
+      if bodyid == 0 or body_weldid[bodyid] == 0:
+        body_invweight0_out[body_invweight0_id, bodyid] = wp.vec2(0.0, 0.0)
+        return
+
+      # Average diagonal: trans = (A[0,0]+A[1,1]+A[2,2])/3, rot = (A[3,3]+A[4,4]+A[5,5])/3
+      inv_trans = (
+        body_A_diag_in[body_A_diag_id, bodyid, 0]
+        + body_A_diag_in[body_A_diag_id, bodyid, 1]
+        + body_A_diag_in[body_A_diag_id, bodyid, 2]
+      ) / 3.0
+      inv_rot = (
+        body_A_diag_in[body_A_diag_id, bodyid, 3]
+        + body_A_diag_in[body_A_diag_id, bodyid, 4]
+        + body_A_diag_in[body_A_diag_id, bodyid, 5]
+      ) / 3.0
+
+      body_invweight0_out[body_invweight0_id, bodyid] = wp.vec2(inv_trans, inv_rot)
+
+    for bodyid in range(1, m.nbody):
+      for row_idx in range(6):
+        wp.launch(
+          compute_body_jac_row,
+          dim=d.nworld,
+          inputs=[
+            m.nv,
+            bodyid,
+            row_idx,
+            m.body_parentid,
+            m.body_rootid,
+            m.body_dofadr,
+            m.body_dofnum,
+            m.dof_parentid,
+            d.subtree_com,
+            d.xipos,
+            d.cdof,
+          ],
+          outputs=[body_jac_row],
+        )
+        smooth.solve_m(m, d, body_result_vec, body_jac_row)
+        wp.launch(
+          compute_body_A_diag_entry,
+          dim=d.nworld,
+          inputs=[m.nv, bodyid, row_idx, body_jac_row, body_result_vec],
+          outputs=[body_A_diag],
+        )
+
+    wp.launch(
+      finalize_body_invweight0,
+      dim=(d.nworld, m.nbody),
+      inputs=[m.body_weldid, body_A_diag],
+      outputs=[m.body_invweight0],
+    )
+  else:
+
+    @nested_kernel(module="unique", enable_backward=False)
+    def zero_body_invweight0(
+      body_invweight0_out: wp.array2d(dtype=wp.vec2),
+    ):
+      worldid, bodyid = wp.tid()
+      body_invweight0_id = worldid % body_invweight0_out.shape[0]
+      body_invweight0_out[body_invweight0_id, bodyid] = wp.vec2(0.0, 0.0)
+
+    wp.launch(
+      zero_body_invweight0,
+      dim=(d.nworld, m.nbody),
+      inputs=[],
+      outputs=[m.body_invweight0],
+    )
+
+  # tendon_invweight0[t] = J_t * inv(M) * J_t'
+  if m.ntendon > 0:
+    ten_J_vec = wp.zeros((d.nworld, m.nv), dtype=float)
```

Comment:
if all the elements in these arrays are computed by the kernels below it should be a bit more performance to utilize `wp.empty` instead of `wp.zeros`